File: haskell/2022/day_3/main.hs
module Day3 where

import Data.List

checkRuck :: [String] -> Int
checkRuck = sum . map (calValue . findChar . splitHalf)
  where
    splitHalf str = [take (length str `div` 2) str, drop (length str `div` 2) str]

checkRucks :: [String] -> Int
checkRucks [] = 0
checkRucks (x : y : z : rest) = calValue (findChar [x, y, z]) + checkRucks rest

calValue :: [Char] -> Int
calValue = sum . map (\c -> maybe 0 (+ 1) (elemIndex c alphabet))

findChar :: [String] -> [Char]
findChar strings
  | length strings == 2 = nub (filter (`elem` last strings) (head strings))
  | length strings == 3 = nub (filter (\c -> c `elem` (strings !! 1) && c `elem` last strings) (head strings))
  | otherwise = []

alphabet :: String
alphabet = ['a' .. 'z'] ++ ['A' .. 'Z']

main :: IO ()
main = do
  file <- readFile "input.txt"
  putStrLn $ "Part 1: " ++ show (checkRuck (lines file))
  putStrLn $ "Part 2: " ++ show (checkRucks (lines file))

File: python/2022/day_3/main.py
def checkRuck(str):
    return sum(map(lambda c: calcValue(findChar(c)), map(splitHalf, str)))


def checkRucks(str):
    if len(str) < 3:
        return 0
    return calcValue(findChar(str[:3])) + checkRucks(str[3:])


def splitHalf(str):
    middle = len(str) // 2
    return [str[:middle], str[middle:]]


def findChar(strings):
    if len(strings) == 2:
        return list(set(strings[0]) & set(strings[1]))
    if len(strings) == 3:
        return list(set(strings[0]) & set(strings[1]) & set(strings[2]))


alphabet = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"


def calcValue(chars): return sum(map(lambda c: alphabet.index(c) + 1, chars))


def part1(file): return checkRuck(read_lines(file))


def part2(file): return checkRucks(read_lines(file))


def read_lines(filename):
    with open(filename) as file:
        return [line.strip() for line in file]
